pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go

```diff
@@ -57,6 +58,14 @@ import (
 
 const workqueueName = "cloudCIDRAllocator"
 
+type stackType int32
+
+const (
+	stackIPv4 stackType = iota
+	stackDualStack
```

Comment:
dualStack can be ipv4-ipv6 or ipv6-ipv4, we need to be completely sure the second will never be enabled, and it we are sure add a comment explaining it

Reply:
The `stackType` type has a slightly different semantics in here. It communicates the cluster stack type, in the following way:

* `stackIPv4`: Pods on the cluster have only IPv4 addresses.
* `stackDualStack`: Pods on the cluster have both IPv4 and IPv6 addresses.
* `stackIPv6`: Pods on the cluster have only IPv6 addresses.

So, `stackDualStack` corresponds to both ipv4-ipv6 and ipv6-ipv4. This is because the `stackType` is only used internally within this file and it is enough to recognize one of the cases ipv4-ipv6/ipv6-ipv4 occurs. At the end we will be taking the corresponding CIDRs from the network interface - either the IPv4 one, or the IPv6 one, or both. The order does not matter, I believe. I think we do not have to explicitly recognize which IP family is the primary for the purposes of configuring Pod CIDRs.

WDYT?

Reply:
Thank you for an explanation offline: it is important to treat the order of service CIDRs as meaningful, because then switching their order at some point in the future will let the IPAM controller automatically handle the right way both an IPv4_IPv6 and an IPv6_IPv4 cluster - even though currently dual stack clusters in GKE are always IPv4_IPv6. I extended the enum to represent the two cases independently.